text/000-structural-and-deref.md

```diff
@@ -0,0 +1,410 @@
+}
+```
+
+we would then bind this in Rust with:
+
+```rust
+#[wasm_bindgen]
+extern {
+    type Parent;
+    #[wasm_bindgen(constructor)]
+    fn new() -> Parent;
+    #[wasm_bindgen(method)]
+    fn method(this: &Parent);
+
+    #[wasm_bindgen(extends = Parent)]
+    type Child;
+    #[wasm_bindgen(constructor)]
+    fn new() -> Child;
+    #[wasm_bindgen(method)]
+    fn method(this: &Child);
+}
+```
+
+and we could then use it like so:
+
+```rust
+#[wasm_bindgen]
+pub fn run() {
+    let parent = Parent::new();
+    parent.method();
+    let child = Child::new();
+    child.method();
+}
+```
+
+and we would today see `parent` and `child` logged to the console. Ok everything
+is working as expected so far! We know we've got `Deref<Target=Parent> for
+Child`, though, so let's say we tweak this example a bit:
+
+```rust
+#[wasm_bindgen]
+pub fn run() {
+    call_method(&Parent::new());
+    call_method(&Child::new());
+}
+
+fn call_method(object: &Parent) {
+    object.method();
+}
+```
+
+Here we'd naively (and correctly) expect `parent` and `child` to be output like
+before, but much to our surprise this actually prints out `parent` twice!
+
+The issue with this is how `#[wasm_bindgen]` treats method calls today. When you
+say:
+
+```rust
+#[wasm_bindgen(method)]
+fn method(this: &Parent);
+```
+
+then `wasm-bindgen` (the CLI tool) generates JS that looks like this:
+
+```js
+const Parent_method_target = Parent.prototype.method;
+
+export function __wasm_bindgen_Parent_method(obj) {
+    Parent_method_target.call(getObject(obj));
+}
+```
+
+Here we can see that, by default, `wasm-bindgen` is **reaching into the
+`prototype` of each class to figure out what method to call**. This in turn
+means that when `Parent::method` is called in Rust, it unconditionally uses the
+method defined on `Parent` rather than walking the protype chain (that JS
+usually does) to find the right `method` method.
+
+To improve the situation there's a `structural` attribute to wasm-bindgen to fix
+this, which when applied like so:
+
+```rust
+#[wasm_bindgen(method, structural)]
+fn method(this: &Parent);
+```
+
+means that the following JS code is generated:
+
+```js
+const Parent_method_target = function() { this.method(); };
+
+// ...
+```
+
+Here we can see that a JS function shim is generated instead of using the raw
+function value in the prototype. This, however, means that our example above
+will indeed print `parent` and then `child` because JS is using prototype
+lookups to find the `method` method.
+
+Phew! Ok with all that information, we can see that **if `structural` is omitted
+then JS class hierarchies can be subtly incorrect when methods taking parent
+classes are passed child classes which override methods**.
+
+An easy solution to this problem is to simply use `structural` everywhere, so...
+let's propose that! Consequently, this RFC proposes changing `#[wasm_bindgen]`
+to act as if all bindings are labeled as `structural`. This will not be a
```

Comment:
Technically this is a breaking change, since somebody might be relying upon the non-structural behavior.

I think that's very unlikely though, so I'm personally fine with not treating it as a breaking change.